tests/unit/utils/test_http.py

```diff
@@ -28,4 +39,19 @@ def bad_iter_content(self, *args, **kwargs):
     with open_url((http / "sample.txt").url) as fd:
         # Test various .read() variants
         assert fd.read(len(text)) == text
-        assert fd.read() == text
+        assert read_nbytes(fd, mode="r")
```

Comment:
As you know under the hood it uses socket read, by definition any io.read it can return data less than expected, even if data is present in the source.
I faced that issue couple of times working in this issue.

In order to have relayable tests you need to have buffer function that would read exact amount of data you requested.

Reply:
Hmm. `fd.read()` can't return anything less than everything. If it does so it's an error. The test suggest it used to work, did it stop?

Reply:
[https://docs.python.org/3/library/io.html](https://docs.python.org/3/library/io.html)

> read(size=-1)
> Read up to size bytes from the object and return them. As a convenience, if size is unspecified or -1, all bytes until EOF are returned. Otherwise, **only one system call is ever made. Fewer than size bytes may be returned if the operating system call returns fewer than size bytes**

As I said before it was failing occasionally. If you are ok with flaky test, let me know i will revert this chanage.

Reply:
> As a convenience, if size is unspecified or -1, **all bytes until EOF** are returned.

You added `read = read1`, which might break it. So, need to check if removing `read = read1` fixes this, otherwise it must be some older bug and I'll create a new issue for `.read()` misbehaving.

And  that read assignment should be removed in either case. `RawIOBase` will take care of it.